contracts/solana/programs/cluster-connection/src/helper.rs

```diff
@@ -0,0 +1,199 @@
+    from: &AccountInfo<'info>,
+    to: &AccountInfo<'info>,
+    system_program: &AccountInfo<'info>,
+    amount: u64,
+) -> Result<()> {
+    let ix = system_instruction::transfer(&from.key(), &to.key(), amount);
+    invoke(
+        &ix,
+        &[from.to_owned(), to.to_owned(), system_program.to_owned()],
+    )?;
+
+    Ok(())
+}
+
+pub fn get_instruction_data(ix_name: &str, data: Vec<u8>) -> Vec<u8> {
+    let preimage = format!("{}:{}", "global", ix_name);
+
+    let mut ix_discriminator = [0u8; 8];
+    ix_discriminator.copy_from_slice(&hash::hash(preimage.as_bytes()).to_bytes()[..8]);
+
+    let mut ix_data = Vec::new();
+    ix_data.extend_from_slice(&ix_discriminator);
+    ix_data.extend_from_slice(&data);
+
+    ix_data
+}
+
+pub fn get_message_hash(from_nid: &String, connection_sn: &u128, message: &Vec<u8>, dst_nid: &String) -> [u8; 32] {
+    let mut encoded_bytes = Vec::new();
+    encoded_bytes.extend(from_nid.as_bytes());
+    encoded_bytes.extend(connection_sn.to_string().as_bytes());
+    encoded_bytes.extend(message);
+    encoded_bytes.extend(dst_nid.as_bytes());
+
+    let hash = hashv(&[&encoded_bytes]);
+
+    hash.to_bytes()
+}
+
+pub fn recover_pubkey(message: [u8; 32], sig: [u8; 65]) -> [u8; 64] {
+    let recovery_id = sig[64] % 27;
+    let signature = &sig[0..64];
+    let recovered_pubkey = secp256k1_recover(&message, recovery_id, signature).unwrap();
+    recovered_pubkey.to_bytes()
+}   
+
+pub fn get_nid(xcall_config: &AccountInfo, config: &Config) -> String {
+    let ix_data = get_instruction_data(GET_NETWORK_ADDRESS, vec![]);
+    let account_metas = vec![AccountMeta::new(xcall_config.key(), false)];
+    let account_infos = vec![xcall_config.to_account_info()];
+
+    let ix = Instruction {
+        program_id: config.xcall,
+        accounts: account_metas,
+        data: ix_data,
+    };
+
+    invoke(&ix, &account_infos).unwrap();
+
+    let network_address = String::from_utf8(get_return_data().unwrap().1).unwrap();
+    network_address.parse::<NetworkAddress>().unwrap().nid()   
+}
+
+pub fn call_xcall_handle_message_with_signatures<'info>(
+    ctx: Context<'_, '_, '_, 'info, ReceiveMessageWithSignatures<'info>>,
+    from_nid: String,
+    message: Vec<u8>,
+    conn_sn: u128,
+    sequence_no: u128,
+    signatures: Vec<[u8; 65]>,
+) -> Result<()> {
+    let mut data = vec![];
+    let dst_nid = get_nid(&ctx.remaining_accounts[1], &ctx.accounts.config);
+    let message_hash = get_message_hash(&from_nid, &conn_sn, &message, &dst_nid);
+    let mut unique_validators = Vec::new();
+    for sig in signatures {
+        let pubkey = recover_pubkey(message_hash, sig);
+        if !unique_validators.contains(&pubkey) && ctx.accounts.config.is_validator(&pubkey) {
+            unique_validators.push(pubkey);
+        } 
+    }
+
+    if (unique_validators.len() as u8) < ctx.accounts.config.threshold {
+        return Err(ConnectionError::ValidatorsMustBeGreaterThanThreshold.into());
+    }
+
```

Comment:
would not it be better to return early when threshold is met  do we need to go through all the signatures?

Reply:
yeah, fixed